src/term/transform/encode_pattern_matching.rs

```diff
@@ -180,6 +193,61 @@ fn make_leaf_pattern_matching_case(
   add_case_to_book(book, crnt_name.clone(), term);
 }
 
+fn make_num_pattern_matching_case(
+  book: &mut Book,
+  def_type: &[Type],
+  def_id: DefId,
+  crnt_name: &Name,
+  crnt_rules: Vec<usize>,
+  match_path: Vec<Pattern>,
+) {
+  use MatchNum::*;
+
+  fn filter_rules(def_rules: &[Rule], crnt_rules: &[usize], arg_idx: usize, Name(ctr): &Name) -> Vec<usize> {
+    crnt_rules
+      .iter()
+      .copied()
+      .filter(|&rule_idx| match &def_rules[rule_idx].pats[arg_idx] {
+        Pattern::Var(_) => false,
+        Pattern::Ctr(_nam, _) => false,
+        Pattern::Num(Succ(_)) => ctr == "+",
+        Pattern::Num(Zero) => ctr == "0",
+        Pattern::Tup(..) => false,
+      })
+      .collect()
+  }
+
+  let make_next_fn_name = |crnt_name, ctr_name: &Name| Name(format!("{crnt_name}$P{ctr_name}"));
+
+  for (next_ctr, next_ctr_args) in vec![(Name::new("0"), vec![]), (Name::new("+"), vec![Name::new("p")])] {
+    let def = &book.defs[&def_id];
+    let crnt_name = make_next_fn_name(crnt_name, &next_ctr);
+    let crnt_rules = filter_rules(&def.rules, &crnt_rules, match_path.len(), &next_ctr);
+    let new_vars =
+      Pattern::Ctr(next_ctr.clone(), vec![Pattern::Var(Some(Name::new(""))); next_ctr_args.len()]);
+    let mut match_path = match_path.clone();
+    match_path.push(new_vars);
+    make_pattern_matching_case(book, def_type, def_id, &crnt_name, crnt_rules, match_path);
+  }
+
+  let term = Term::Var { nam: Name::new("x").clone() };
+  let zero_id = book.def_names.def_id(&make_next_fn_name(crnt_name, &Name::new("0"))).unwrap();
+  let succ_id = book.def_names.def_id(&make_next_fn_name(crnt_name, &Name::new("+"))).unwrap();
+
+  let term = Term::Match {
+    scrutinee: Box::new(term),
+    arms: vec![
+      (Pattern::Num(MatchNum::Zero), Term::Ref { def_id: zero_id }),
+      (Pattern::Num(MatchNum::Zero), Term::Ref { def_id: succ_id }),
```

Comment:
MatchNum::Succ

Reply:
The pattern for the second is not Zero, otherwise youre creating this
```
match x {
 0: Foo$P0
 0: Foo$P+
}
```

Reply:
I see, but this looks rather wrong, so maybe we should find a better way of doing this in a future PR.